lib/buffy/throttle_and_timed.ex

```diff
@@ -0,0 +1,372 @@
+
+        defmodule MyThrottler do
+          use Buffy.ThrottleAndTimed,
+            registry_module: Horde.Registry,
+            registry_name: MyApp.HordeRegistry,
+            supervisor_module: Horde.DynamicSupervisor,
+            supervisor_name: MyApp.HordeDynamicSupervisor,
+            throttle: :timer.minutes(2),
+            loop_timeout: :timer.minutes(10)
+
+          def handle_throttle(args) do
+            # Do something with args
+          end
+        end
+
+  ## Telemetry
+
+  These are the events that are called by the `Buffy.ThrottleAndTimed` module:
+
+  - `[:buffy, :throttle, :throttle]` - Emitted when the `throttle/1` function is called.
+  - `[:buffy, :throttle, :timeout]` - Emitted when inbox timeout is triggered.
+  - `[:buffy, :throttle, :handle, :start]` - Emitted at the start of the `handle_throttle/1` function.
+  - `[:buffy, :throttle, :handle, :stop]` - Emitted at the end of the `handle_throttle/1` function.
+  - `[:buffy, :throttle, :handle, :exception]` - Emitted when an error is raised in the `handle_throttle/1` function.
+
+  All of these events will have the following metadata:
+
+  - `:args` - The arguments passed to the `throttle/1` function.
+  - `:key` - A hash of the passed arguments used to deduplicate the throttled function.
+  - `:module` - The module using `Buffy.ThrottleAndTimed`.
+
+  With the additional metadata for `[:buffy, :throttle, :handle, :stop]`:
+
+  - `:result` - The return value of the `handle_throttle/1` function.
+
+  """
+  require Logger
+  alias Buffy.ThrottleAndTimed
+
+  @typedoc """
+  A list of arbitrary arguments that are used for the `c:handle_throttle/1`
+  function.
+  """
+  @type args :: term()
+
+  @typedoc """
+  A unique key for debouncing. This is used for GenServer uniqueness and is
+  generated from hashing all of the args.
+  """
+  @type key :: term()
+
+  @typedoc """
+  Internal state that `Buffy.ThrottleAndTimed` keeps.
+  """
+  @type state :: %{
+          key: key(),
+          args: args(),
+          timer_ref: reference()
+        }
+
+  @doc """
+  A function to call the throttle. This will start
+  and wait the configured `throttle` time before calling the `c:handle_throttle/1`
+  function.
+  """
+  @callback throttle(args :: args()) :: :ok | {:error, term()}
+
+  @doc """
+  The function called after the throttle has completed. This function will
+  receive the arguments passed to the `throttle/1` function.
+  """
+  @callback handle_throttle(args()) :: any()
+
+  defmacro __using__(opts) do
+    registry_module = Keyword.get(opts, :registry_module, Registry)
+    registry_name = Keyword.get(opts, :registry_name, Buffy.Registry)
+    restart = Keyword.get(opts, :restart, :temporary)
+    supervisor_module = Keyword.get(opts, :supervisor_module, DynamicSupervisor)
+    supervisor_name = Keyword.get(opts, :supervisor_name, Buffy.DynamicSupervisor)
+    throttle = Keyword.fetch!(opts, :throttle)
```

Comment:
Should we apply the same validation to make sure `throttle` is a number?

Reply:
I think it would crash pretty fast because that's used during `init` - therefore we'd already be getting the necessary feedback that we want as-is.